Allow for Union-typed memory aliasing

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58911
* #58673
* #57014
* #57013
* __->__ #57012
* #57011
* #53180

Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

Commit:
Allow for Union-typed memory aliasing

[ghstack-poisoned]

Commit:
Update on "Allow for Union-typed memory aliasing"




[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

eellison commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -41,41 +67,75 @@ class MutableTypePtrHelper {
+        }));
+  }
+
   c10::optional<TypePtr> getMutableTypeImpl(const TypePtr& type) {
```
In the previous implementation, we were returning a vector of types here? why was that changed ?

ansley replied:
In the current implementation, the signature is the same as it is here--`c10::optional<TypePtr> getMutableTypeImpl(const TypePtr& type)`. By not changing the signature, I was able to keep more of the original logic. Also, in most cases, we won't be returning multiple elements in the vector. Wouldn't returning a vector just cause lots of extra `if (vec.size() == ...` checks?

eellison replied:
I think some of those changes are needed, like here https://github.com/pytorch/pytorch/blob/6dea82a8edce728b8c4850b884bbb976a1a59c18/torch/csrc/jit/ir/alias_analysis.cpp#L1153. 

If we refine a `Union[List[int], Tensor]` to `Tensor` we want the refined Tensor Value* to point to the Union Value*, but that won't occur right now because they will be returning different Mutable Types. 

`getMutableType` could have always been `mapTypeToTypeSetWhichItCanAlias`, however because the only form of Unions previously we're `Union[T, None]` we always knew that the typeSet was size 1

ansley replied:
Resolved offline: my current implementation isn't wrong, but we will change `getMutableType` to return a `std::vector<TypePtr>` since that will make the API more intuitive

eellison reviewed (commented):
The last version of this I reviewed looked pretty close, but this looks very different now. Could we bring the implementation closer to how it was ? In particular, getMutableType should be returning a vector of mutable types as it was in the last version

ansley commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -12,25 +12,51 @@ namespace jit {
 
+  // Map any mutable type to a type such that all other types which the
+  // mutable type can alias will be mapped to the same type. This
+  // follows a similar logic to `unifyTypes` because any two mutable
```
Same semantics, but imo the original wording could be improved. I had to read that first sentence several times to picture it in my head. I don't mind keeping the original if I'm the only one who has problems with it, though :)

eellison replied:
The semantics are a bit different, as the `Tuple[None, Tensor]` and `Tuple[List[int], None]` case show

ansley replied:
Is the issue here with "This function follows a similar logic to `unifyTypes` because..."? That line is from the original wording. It was confusing to me when I was trying to understand the alias analysis pass (and it sounds like you're saying it's downright misleading), so maybe we should take it out altogether

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

ansley commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1658,20 +1658,21 @@ struct to_ir {
       SugaredValuePtr false_val = save_false->findInParentFrame(x);
-      bool default_to_union = false;
       if (true_val && false_val) {
         if (tv->type() == fv->type() && tv->type()->kind() == UnionType::Kind) {
```
Fixed in previous PR

ansley commented on test/jit/test_union.py:
```diff
@@ -487,3 +487,25 @@ def script(x: Union[List[int], int]) -> int:
+                x_alias.append(torch.tensor(3))
+            return x
+
+        self.assertEqual(eager(), script())
```
I think this was from when I didn't understand `torch.jit.isinstance` 🥲

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

eellison commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -13,70 +13,104 @@ namespace jit {
-        } else {
-          return TupleType::create(mutable_types);
         }
+        return c10::optional<TypePtr>(TupleType::create(mutable_types));
```
nit: you don't need to wrap it in `c10::optional<TypePtr>`

eellison reviewed (commented):
Would it be possible to separate out the non-semantic changes? I think 80% of this review is non-semantic. I think separating that out would make iterating on making changes and reviewing much quicker, because the non-semantic changes do not require as close attention for reviewing.

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

ansley commented on torch/csrc/jit/passes/utils/memory_dag.cpp:
```diff
@@ -193,20 +197,22 @@ void MemoryDAG::setWildcards(
       continue;
     }
 
-    auto wildcardElement = getWildcardElement(*(e->values.begin()));
```
The name of the variable has been changed from `wildcardElement` to `wildcardElement_List` to denote that it is no longer a single element but rather a list. I'm reverting the rest of the changes

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

eellison reviewed (commented):
Looks great !! Couple of small comments, we discussed offline the 1-1 mapping of union wildcards to elements i think with that this is ready to go

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

eellison commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -1221,7 +1267,11 @@ void AliasDb::giveFreshAlias(
   if (add_wildcard_to_contained_elems) {
-    addContainedTypesToFreshElement(new_elem, *maybe_mut_type);
+    if ((*maybe_mut_types).size() > 1) {
+      pointUnionTypeElementToAllContainedTypes(new_elem, *maybe_mut_types);
```
nit: I would add an assertion that the type is a Union type here. I dont know if we have Optional[Union] right now (eek) so maybe this is a todo after we remove optionals

ansley replied:
It's not possible to have `Optional[Union]`! I changed that in the first PR to always be `Union[None, ...]`

eellison commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -1207,8 +1253,8 @@ void AliasDb::createValue(const Value* value) {
 void AliasDb::giveFreshAlias(
```
`add_wildcard_to_contained_elems` is a micro-optimization for nodes like 

```
x = torch.tensor([2])
torch.cat([x, x, x, x])
```
Where we look if the list created is passed to a functional operator and then put x in the contained elements of the list instead of setting x to the wildcard set. Even if this parameter is `False`,  we still want to `pointUnionTypeElementToAllContainedTypes`. It maybe makes sense to rename the argument to something like `contained_elements_have_already_added` or something, i'm not sure.

eellison commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -13,90 +13,134 @@ namespace jit {
-      mutable_type_cache_->emplace(type, *mutable_type);
+    auto mutable_types = mapTypeToAliasTypeSetImpl(type);
+    if (mutable_type_cache_ && mutable_types) {
+      mutable_type_cache_->emplace(type, *mutable_types);
```
nit: maybe add an assertion here that `mutable_types->size() != 0`

eellison reviewed (approved):
Great job 🚀 🚀 🚀  

The one comment about adding `add_wildcard_to_contained_elems` should be easily addressable without another review from me so this is good to go from my end provided the `TORCH_INTERNAL_ASSERT(e->pointsTo.empty())` is fine to remove - I think it should but would like to double check with @suo before landing

ansley commented on torch/csrc/jit/passes/utils/memory_dag.cpp:
```diff
@@ -189,7 +191,6 @@ void MemoryDAG::setWildcards(
   for (const std::unique_ptr<Element>& e : this->indexToElementMap_) {
     if (e->values.empty()) {
       // This element is a wildcard element, we can skip it.
-      TORCH_INTERNAL_ASSERT(e->pointsTo.empty());
```
This is what I want the Union wildcards to look like:

```
WILDCARD(Union[T1, T2, T3]) -----> WILDCARD(T1)
                           |
                            -----> WILDCARD(T2)
                           |
                            -----> WILDCARD(T3)
```

So, the wildcard of a Union should have empty `values`, but it should point to other wildcards, which means that `pointsTo` won't be empty. This was not a possibility before implementing Union, which is why the assert worked

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

Commit:
Update on "Allow for Union-typed memory aliasing"


Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

[ghstack-poisoned]
